bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieImpl.java

```diff
@@ -849,12 +852,13 @@ public void run() {
     public int shutdown() {
         return shutdown(ExitCode.OK);
     }
-
     // internal shutdown method to let shutdown bookie gracefully
     // when encountering exception
-    synchronized int shutdown(int exitCode) {
+    ReentrantLock lock = new ReentrantLock(true);
+    int shutdown(int exitCode) {
```

Comment:
Why use `ReentrantLock` instead of `synchronized` ?

Reply:
> I think the bad case is that if a journal crash first and already triggers BookieImpl.triggerBookieShutdown(), then we shutdown bookie by the command bookkeeper-deamon stop bookie, the shutdown progress by the command may not execute completely. 

This is the reason.